### ISSUE TITLE
Add 'Plugin manager' section to 'napari architecture guide'

### ISSUE THREAD

**Issue by lucyleeow** [napari/docs#321]
## 🧰 Task
We are lacking documentation on the npe2 Plugin manager. The 'plugins' section talks about building plugins, and information about the plugin manager would not be relevant here. Plugin manager information is more relevant to napari developers. Currently there is some mention of it in NAP 2.

E.g., information about the plugin manager command registry and indexing would fit here, see: https://github.com/napari/npe2/issues/314

cc @jaimergp and @DragaDoncila who may have the knowledge to add such a section.

**Comment by DragaDoncila** [comment]
Ahh @lucyleeow when you mentioned this, I was thinking about the [ plugin installation manager](https://github.com/napari/napari-plugin-manager) but I see now that you mean the internal plugin manager. It would probably be nice for us to disambiguate this in the docs also 😆 .

**Comment by DragaDoncila** [comment]
Information that would be good to include:

- how to access the plugin manager during a napari session and check discovered plugins etc.
- how to dynamically register a plugin? maybe?
- where/when the manager is initialized and plugins discovered
- the npe2 shim? related but maybe not directly relevant?

**Comment by lucyleeow** [comment]
> how to dynamically register a plugin? maybe? 

Is this in the context of testing? Either way, I think it would be useful to have a section on testing plugins in napari, Dynamic Plugin and how to add contributions via the `ContributionDecorator`s (and how they work under the hood).

> I was thinking about the[ plugin installation manager](https://github.com/napari/napari-plugin-manager) but I see now that you mean the internal plugin manager.

This has tripped me up too. Naming is hard.

**Comment by lucyleeow** [comment]
Noticed in #370 that we do not have info on `DynamicPlugin`s and how to add plugin contributions with the decorators. I think this task could fall under here. 

(I know its mentioned above in https://github.com/napari/docs/issues/321#issuecomment-1907365595 but just highlighting again)

**Comment by jaimergp** [comment]
> I was thinking about the [plugin installation manager](https://github.com/napari/napari-plugin-manager) but I see now that you mean the internal plugin manager

Oh wow, yes, that's confusing, I had to read that sentence a few times to actually parse it 😬 

- Plugin manager as in the library to handle plugin registration. This is `napari-plugin-engine` and `npe2`, right?
- Plugin (installation) manager as the UI to handle the installation of plugins with conda and pip. This is `napari-plugin-manager`.

Are there better terms we can use? Should we exclusively refer to the plugin registration library as the "plugin engine" instead of manager?

**Comment by lucyleeow** [comment]
Yes there should be different terms!

npe1 and npe2 (the plugin engines) both have a class called `PluginManager` (see: [npe2 PM](https://github.com/napari/npe2/blob/37d4ba4828b0e8680dcb08e7536d1951c9e8b62a/src/npe2/_plugin_manager.py#L223)). I think this keeps a registry of plugins, their status etc.

Maybe Plugin installation manager GUI, could be "plugin manager widget"? or "plugin installation manager"?

cc @DragaDoncila

**Comment by DragaDoncila** [comment]
Mmm yeah this is a sad state of affairs 😆 . I think we should rename the internal plugin manager(s) rather than the plugin installation manager, since that is a public facing package and also is named in the menus, while the internal plugin managers aren't. I would be in favour of renaming to something like `PluginRegistry` for the internal one(s). Would probably want to do this post npe1 deprecation I guess?

**Comment by lucyleeow** [comment]
> I think we should rename the internal plugin manager(s) rather than the plugin installation manager, since that is a public facing package and also is named in the menus

Yeah good point. I just don't think of the GUI when I hear 'plugin manager' and then have to remind myself it's the GUI thing.

> something like PluginRegistry for the internal one

Any potential with confusion that this would be an app-model thing because app-model has the CommandRegistry, MenuRegistry...etc ?

**Comment by psobolewskiPhD** [comment]
What a can of worms...

the plugin manager used to install/etc. plugins isn't live yet--0.4.19 uses the old UI--so in principle we could still change that if it would make life easier...
